File: android/app/src/main/java/com/scanx/app/convert/PhotoTranslateRenderer.kt
package com.scanx.app.convert

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Vẽ bản dịch ĐÈ lên ảnh gốc, kiểu Google Dịch (bản 1.0): mỗi khối chữ được phủ bằng màu nền đo quanh
 * khối, rồi viết bản dịch vào đúng khung đó (tự xuống dòng, tự thu nhỏ cỡ chữ cho vừa), màu chữ lấy
 * theo màu mực gốc nếu đủ tương phản, không thì đen/trắng tuỳ nền.
 */
object PhotoTranslateRenderer {

    /** [translations] cùng thứ tự với [blocks]; phần tử null = giữ nguyên chữ gốc ở khối đó. */
    fun render(source: Bitmap, blocks: List<PhotoTranslation.TextBlock>, translations: List<String?>): Bitmap {
        val out = source.copy(Bitmap.Config.ARGB_8888, true)
        val canvas = Canvas(out)
        val fill = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
        for ((i, block) in blocks.withIndex()) {
            val text = translations.getOrNull(i)?.trim().orEmpty()
            if (text.isEmpty()) continue
            val pad = max(2f, block.lineHeight * 0.12f)
            val rect = RectF(
                (block.box.left - pad).coerceAtLeast(0f),
                (block.box.top - pad).coerceAtLeast(0f),
                (block.box.right + pad).coerceAtMost(out.width.toFloat()),
                (block.box.bottom + pad).coerceAtMost(out.height.toFloat()),
            )
            if (rect.width() < 4f || rect.height() < 4f) continue
            val bg = borderColor(source, rect)
            val ink = inkColor(source, rect, bg)
            fill.color = bg
            canvas.drawRoundRect(rect, pad, pad, fill)
            drawFitted(canvas, text, rect, ink, block.lineHeight)
        }
        return out
    }

    private fun drawFitted(canvas: Canvas, text: String, rect: RectF, color: Int, lineHeight: Float) {
        val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            this.color = color
            typeface = Typeface.DEFAULT
        }
        val width = rect.width().roundToInt().coerceAtLeast(1)
        var size = (lineHeight * 0.78f).coerceIn(8f, 160f)
        var layout = build(text, paint, width, size)
        while (layout.height > rect.height() && size > 8f) {
            size *= 0.9f
            layout = build(text, paint, width, size)
        }
        canvas.save()
        canvas.clipRect(rect)
        val dy = ((rect.height() - layout.height) / 2f).coerceAtLeast(0f)
        canvas.translate(rect.left, rect.top + dy)
        layout.draw(canvas)
        canvas.restore()
    }

    private fun build(text: String, paint: TextPaint, width: Int, size: Float): StaticLayout {
        paint.textSize = size
        return StaticLayout.Builder.obtain(text, 0, text.length, paint, width)
            .setAlignment(Layout.Alignment.ALIGN_NORMAL)
            .setLineSpacing(0f, 1f)
            .setIncludePad(false)
            .build()
    }

    /** Màu nền = trung vị từng kênh của các điểm trên viền ngoài khối. */
    private fun borderColor(bmp: Bitmap, r: RectF): Int {
        val rs = ArrayList<Int>(); val gs = ArrayList<Int>(); val bs = ArrayList<Int>()
        fun sample(x: Float, y: Float) {
            val xi = x.roundToInt().coerceIn(0, bmp.width - 1)
            val yi = y.roundToInt().coerceIn(0, bmp.height - 1)
            val c = bmp.getPixel(xi, yi)
            rs.add(Color.red(c)); gs.add(Color.green(c)); bs.add(Color.blue(c))
        }
        val n = 24
        for (k in 0..n) {
            val x = r.left + r.width() * k / n
            val y = r.top + r.height() * k / n
            sample(x, r.top - 2f); sample(x, r.bottom + 2f)
            sample(r.left - 2f, y); sample(r.right + 2f, y)
        }
        return Color.rgb(median(rs), median(gs), median(bs))
    }

    /** Màu mực: trung vị các điểm trong khối khác nền rõ rệt; không đủ tương phản → đen/trắng theo nền. */
    private fun inkColor(bmp: Bitmap, r: RectF, bg: Int): Int {
        val bgL = luminance(bg)
        val step = max(1f, min(r.width(), r.height()) / 20f)
        val rs = ArrayList<Int>(); val gs = ArrayList<Int>(); val bs = ArrayList<Int>()
        var y = r.top
        while (y < r.bottom) {
            var x = r.left
            while (x < r.right) {
                val c = bmp.getPixel(x.roundToInt().coerceIn(0, bmp.width - 1), y.roundToInt().coerceIn(0, bmp.height - 1))
                if (kotlin.math.abs(luminance(c) - bgL) > 0.35f) { rs.add(Color.red(c)); gs.add(Color.green(c)); bs.add(Color.blue(c)) }
                x += step
            }
            y += step
        }
        val fallback = if (bgL > 0.5f) Color.rgb(20, 20, 20) else Color.WHITE
        if (rs.size < 8) return fallback
        val ink = Color.rgb(median(rs), median(gs), median(bs))
        return if (kotlin.math.abs(luminance(ink) - bgL) > 0.4f) ink else fallback
    }

    private fun luminance(c: Int) = (0.299f * Color.red(c) + 0.587f * Color.green(c) + 0.114f * Color.blue(c)) / 255f

    private fun median(v: MutableList<Int>): Int {
        if (v.isEmpty()) return 255
        v.sort()
        return v[v.size / 2]
    }
}

File: ios/ScanX/Translate/PhotoTranslateRenderer.swift
import UIKit

/// Vẽ bản dịch ĐÈ lên ảnh kiểu Google Dịch — port từ Android PhotoTranslateRenderer.kt: phủ khối bằng màu nền
/// đo quanh khối, viết bản dịch vào đúng khung (tự xuống dòng, tự thu nhỏ, không to hơn chữ gốc), màu chữ
/// theo mực gốc nếu đủ tương phản; vẽ theo góc nghiêng của chữ (bản 1.1).
enum PhotoTranslateRenderer {
    struct Colors {
        let background: UIColor
        let ink: UIColor
    }

    /// translations cùng thứ tự với blocks; nil = giữ nguyên chữ gốc ở khối đó.
    static func render(_ image: CGImage, blocks: [PhotoTranslation.TextBlock], translations: [String?]) -> UIImage? {
        guard let pixels = ScanFilters.rgbxBuffer(from: image) else { return nil }
        let size = CGSize(width: image.width, height: image.height)
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = true
        return UIGraphicsImageRenderer(size: size, format: format).image { ctx in
            UIImage(cgImage: image).draw(in: CGRect(origin: .zero, size: size))
            for (i, block) in blocks.enumerated() {
                guard i < translations.count,
                      let text = translations[i]?.trimmingCharacters(in: .whitespacesAndNewlines),
                      !text.isEmpty
                else { continue }
                drawBlock(ctx.cgContext, block: block, text: text, colors: blockColors(pixels, block))
            }
        }
    }

    private static func pad(_ block: PhotoTranslation.TextBlock) -> Double {
        max(2, block.lineHeight * 0.12)
    }

    private static func rect(_ block: PhotoTranslation.TextBlock) -> CGRect {
        let p = pad(block)
        return CGRect(
            x: block.box.left - p,
            y: block.box.top - p,
            width: block.box.width + 2 * p,
            height: block.box.height + 2 * p
        )
    }

    /// Phủ khối và viết bản dịch (toạ độ ảnh; ngữ cảnh UIKit trục y hướng xuống).
    static func drawBlock(_ cg: CGContext, block: PhotoTranslation.TextBlock, text: String, colors: Colors) {
        let r = rect(block)
        if r.width < 4 || r.height < 4 { return }
        let p = CGFloat(pad(block))
        UIGraphicsPushContext(cg)
        cg.saveGState()
        if block.angle != 0 { cg.rotate(by: CGFloat(block.angle * .pi / 180)) }
        colors.background.setFill()
        UIBezierPath(roundedRect: r, cornerRadius: p).fill()
        drawFitted(cg, text: text, rect: r, color: colors.ink, lineHeight: block.lineHeight)
        cg.restoreGState()
        UIGraphicsPopContext()
    }

    private static func drawFitted(_ cg: CGContext, text: String, rect: CGRect, color: UIColor, lineHeight: Double) {
        // Cỡ chữ gốc ≈ 0,8 × chiều cao dòng thật; bản dịch không to hơn, chỉ thu nhỏ cho vừa khung.
        var size = CGFloat(min(max(lineHeight * 0.8, 6), 160))
        func attributed(_ s: CGFloat) -> NSAttributedString {
            let style = NSMutableParagraphStyle()
            style.lineBreakMode = .byWordWrapping
            return NSAttributedString(string: text, attributes: [
                .font: UIFont.systemFont(ofSize: s),
                .foregroundColor: color,
                .paragraphStyle: style,
            ])
        }
        func height(_ a: NSAttributedString) -> CGFloat {
            a.boundingRect(
                with: CGSize(width: rect.width, height: .greatestFiniteMagnitude),
                options: [.usesLineFragmentOrigin, .usesFontLeading],
                context: nil
            ).height.rounded(.up)
        }
        var attr = attributed(size)
        var h = height(attr)
        while h > rect.height && size > 6 {
            size *= 0.92
            attr = attributed(size)
            h = height(attr)
        }
        cg.saveGState()
        cg.clip(to: rect)
        let dy = max(0, (rect.height - h) / 2)
        attr.draw(with: CGRect(x: rect.minX, y: rect.minY + dy, width: rect.width, height: h),
                  options: [.usesLineFragmentOrigin, .usesFontLeading], context: nil)
        cg.restoreGState()
    }

    // MARK: - Màu nền / màu mực

    static func blockColors(_ pixels: ScanFilters.RGBX, _ block: PhotoTranslation.TextBlock) -> Colors {
        let r = rect(block)
        let bg = borderColor(pixels, block, r)
        return Colors(background: uiColor(bg), ink: uiColor(inkColor(pixels, block, r, bg)))
    }

    private typealias RGB = (r: Int, g: Int, b: Int)

    private static func pixel(_ pixels: ScanFilters.RGBX, _ block: PhotoTranslation.TextBlock, _ x: Double, _ y: Double) -> RGB {
        let (ix, iy) = block.toImage(x, y)
        let px = clampInt(Int(ix.rounded()), 0, pixels.width - 1)
        let py = clampInt(Int(iy.rounded()), 0, pixels.height - 1)
        let i = (py * pixels.width + px) * 4
        return (Int(pixels.pixels[i]), Int(pixels.pixels[i + 1]), Int(pixels.pixels[i + 2]))
    }

    private static func borderColor(_ pixels: ScanFilters.RGBX, _ block: PhotoTranslation.TextBlock, _ r: CGRect) -> RGB {
        var samples: [RGB] = []
        let n = 24
        for k in 0...n {
            let x = Double(r.minX + r.width * CGFloat(k) / CGFloat(n))
            let y = Double(r.minY + r.height * CGFloat(k) / CGFloat(n))
            samples.append(pixel(pixels, block, x, Double(r.minY) - 2))
            samples.append(pixel(pixels, block, x, Double(r.maxY) + 2))
            samples.append(pixel(pixels, block, Double(r.minX) - 2, y))
            samples.append(pixel(pixels, block, Double(r.maxX) + 2, y))
        }
        return medianRGB(samples)
    }

    private static func inkColor(_ pixels: ScanFilters.RGBX, _ block: PhotoTranslation.TextBlock, _ r: CGRect, _ bg: RGB) -> RGB {
        let bgL = luminance(bg)
        let step = max(1, Double(min(r.width, r.height)) / 20)
        var samples: [RGB] = []
        var y = Double(r.minY)
        while y < Double(r.maxY) {
            var x = Double(r.minX)
            while x < Double(r.maxX) {
                let c = pixel(pixels, block, x, y)
                if abs(luminance(c) - bgL) > 0.35 { samples.append(c) }
                x += step
            }
            y += step
        }
        let fallback: RGB = bgL > 0.5 ? (20, 20, 20) : (255, 255, 255)
        if samples.count < 8 { return fallback }
        let ink = medianRGB(samples)
        return abs(luminance(ink) - bgL) > 0.4 ? ink : fallback
    }

    private static func luminance(_ c: RGB) -> Double {
        (0.299 * Double(c.r) + 0.587 * Double(c.g) + 0.114 * Double(c.b)) / 255
    }

    private static func medianRGB(_ v: [RGB]) -> RGB {
        if v.isEmpty { return (255, 255, 255) }
        let rs = v.map { $0.r }.sorted()
        let gs = v.map { $0.g }.sorted()
        let bs = v.map { $0.b }.sorted()
        return (rs[rs.count / 2], gs[gs.count / 2], bs[bs.count / 2])
    }

    private static func uiColor(_ c: RGB) -> UIColor {
        UIColor(red: CGFloat(c.r) / 255, green: CGFloat(c.g) / 255, blue: CGFloat(c.b) / 255, alpha: 1)
    }
}
